src/Providers/Http/DTO/Request.php

```diff
@@ -0,0 +1,371 @@
+        $new = clone $this;
+        $new->headers = $newHeaders;
+        return $new;
+    }
+
+    /**
+     * Returns a new instance with the specified data.
+     *
+     * @since n.e.x.t
+     *
+     * @param string|array<string, mixed> $data The request data.
+     * @return self A new instance with the data.
+     */
+    public function withData($data): self
+    {
+        $new = clone $this;
+        if (is_string($data)) {
+            $new->body = $data;
+            $new->data = null;
+        } elseif (is_array($data)) {
+            $new->data = $data;
+            $new->body = null;
+        } else {
+            $new->data = null;
+            $new->body = null;
+        }
+        return $new;
+    }
+
+    /**
+     * Gets the request data array.
+     *
+     * @since n.e.x.t
+     *
+     * @return array<string, mixed>|null The request data array.
+     */
+    public function getData(): ?array
+    {
+        return $this->data;
+    }
+
+    /**
+     * Gets the request body string.
+     *
+     * @since n.e.x.t
+     *
+     * @return string|null The request body string.
+     */
+    public function getBodyString(): ?string
+    {
+        return $this->body;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     */
+    public static function getJsonSchema(): array
+    {
+        return [
+            'type' => 'object',
+            'properties' => [
+                self::KEY_METHOD => [
+                    'type' => 'string',
+                    'description' => 'The HTTP method.',
+                ],
+                self::KEY_URI => [
+                    'type' => 'string',
+                    'description' => 'The request URI.',
+                ],
+                self::KEY_HEADERS => [
+                    'type' => 'object',
+                    'additionalProperties' => [
+                        'type' => 'array',
+                        'items' => ['type' => 'string'],
+                    ],
+                    'description' => 'The request headers.',
+                ],
+                self::KEY_DATA => [
+                    'type' => ['string', 'array', 'null'],
+                    'description' => 'The request data.',
+                ],
+            ],
+            'required' => [self::KEY_METHOD, self::KEY_URI, self::KEY_HEADERS],
+        ];
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     *
+     * @return RequestArrayShape
+     */
+    public function toArray(): array
+    {
+        $array = [
+            self::KEY_METHOD => $this->method->value,
+            self::KEY_URI => $this->uri,
+            self::KEY_HEADERS => $this->headers->getAll(),
+        ];
+
+        // Include whichever data type is set
+        if ($this->body !== null) {
+            $array[self::KEY_DATA] = $this->body;
+        } elseif ($this->data !== null) {
+            $array[self::KEY_DATA] = $this->data;
+        }
```

Comment:
See my comment above: I think it would be great to expose this with the `$data` part internally handled, i.e. either baked into `KEY_URI` or `KEY_BODY`.

Something like this:
```suggestion
        $array = [
            self::KEY_METHOD => $this->method->value,
            self::KEY_URI => $this->getUri(),
            self::KEY_HEADERS => $this->headers->getAll(),
        ];

        $body = $this->getBody();
        if ($body !== null) {
            $array[self::KEY_BODY] = $body;
        }
```

Reply:
Resolved in adfcd5f